### PR TITLE
Notify slack when we receive suspicious collective data (SPAM)

### PR DESCRIPTION
**Pull request by CanerSezgin**
Issue: https://github.com/opencollective/opencollective/issues/2750

@Betree 
Before going further, couple of more questions. 

1) I am planning to call **getSuspiciousKeywords** before creating collective. You can find example for `server/graphql/v1/mutations/collectives.js `-> **createCollective** function. Is this the correct position ? 

2) Should I call same function for these 3 (only) **createCollective**, **createCollectiveFromGithub**, **editCollective** If this is the case, I am going to create a function for this approach as well. (I guess again better place is spam.ts)
_PS: with longDescription if it is applicable_
```
const suspiciosKeywordCheckList = ['name', 'website', 'description'];
    suspiciosKeywordCheckList.forEach(prop => {
      const suspiciousKeywords = getSuspiciousKeywords(collectiveData[prop]);
      if (suspiciousKeywords.length) {
        console.log('Suspicious Keyword !!!', suspiciousKeywords, prop);
      }
});
```

3) According to the answer of the second question, we may think to create a class rather than some functions in spam.ts (I suppose, more spam business logic will be added later)

4) About slack notification, I don't have any access to **spam-reports** channel I guess. This is why I couldn't try but basically I suppose that I need to call only one function at the same place instead of console log in the example code block. If it is the case could you please send me function for sending notification. (or basically I can leave it to you guys)

### CODE REVIEW

**Comment by Betree** [comment]
> 1. I am planning to call **getSuspiciousKeywords** before creating collective. You can find example for `server/graphql/v1/mutations/collectives.js `-> **createCollective** function. Is this the correct position ?
> 2) Should I call same function for these 3 (only) **createCollective**, **createCollectiveFromGithub**, **editCollective** If this is the case, I am going to create a function for this approach as well. (I guess again better place is spam.ts)
_PS: with longDescription if it is applicable_

To make sure that we cover all the paths, I'd say that the best place to implement would be in the `afterCreate`: https://github.com/opencollective/opencollective-api/blob/edb538ea1161d7033600dacfd02979673a3fa350/server/models/Collective.js#L637

> 3. According to the answer of the second question, we may think to create a class rather than some functions in spam.ts (I suppose, more spam business logic will be added later)

This is up to you, this repo usually tends to use mostly functional code but feel free to use a class if you think that it is the best pattern to implement this.

> 4. About slack notification, I don't have any access to **spam-reports** channel I guess. This is why I couldn't try but basically I suppose that I need to call only one function at the same place instead of console log in the example code block. If it is the case could you please send me function for sending notification. (or basically I can leave it to you guys)

You can eventually try with this (public) channel: https://opencollective.slack.com/archives/C1T3E0T5F

**Comment by Betree** [comment]
You'll need to rebase on master to get the tests working (see https://github.com/opencollective/opencollective-api/pull/3123). Sorry for the inconvenience!

**Comment by CanerSezgin** [comment]
> > 1. Should I call same function for these 3 (only) **createCollective**, **createCollectiveFromGithub**, **editCollective** If this is the case, I am going to create a function for this approach as well. (I guess again better place is spam.ts)
> >    _PS: with longDescription if it is applicable_
> 
> To make sure that we cover all the paths, I'd say that the best place to implement would be in the `afterCreate`:
> 
> https://github.com/opencollective/opencollective-api/blob/edb538ea1161d7033600dacfd02979673a3fa350/server/models/Collective.js#L637
> 

I am just calling spamController as you suggest in **afterCreate**, I put it inside if check but it might be redundant. I don't know what kind of instance we have here but I saw there is a similar logic above this is why created like this

```
if (instance.type === 'COLLECTIVE' || instance.type === 'EVENT') {
            await models.PaymentMethod.create({
                         ...Current Codes...
            });
}

if (instance.type === 'COLLECTIVE') {
      const spamCheck = spamController.collectiveCheck(instance);
}
```

With this approach, when I created collective from frontend, it works fine for  **createCollective**, couldn't try for **createCollectiveFromGithub** but it will be ok most prob as you suggest. 
However when I edit (directly from frontend), I didn't see **afterCreate** has been called. I guess we need to call spamController somewhere else to use with edit. 

Waiting your comments on this @Betree 
Meanwhile, couldn't try to send these messages on slack, will try to implement this. 

For now, I am returning this kind of array as a result, it shows which part contains bannedWords
```
{ 
  collectiveId: 10894,
  warnings: { 
     name: [ 'keto', 'pills' ], 
     description: [ 'keto' ] 
  } 
}
```

**Comment by CanerSezgin** [comment]
Another question, what do you suggest for this error. I am using it in another file but TS wants me to use in the same file that class is defined.
`[error]server/lib/spam.ts(43,13): error TS6133: 'collectiveCheck' is declared but its value is never read.`

**Comment by Betree** [comment]
> However when I edit (directly from frontend), I didn't see afterCreate has been called. I guess we need to call spamController somewhere else to use with edit.

`afterCreate` is only called after creation. You can also try `afterUpdate` or `afterSave` hooks: https://sequelize.org/master/manual/hooks.html

> Another question, what do you suggest for this error. I am using it in another file but TS wants me to use in the same file that class is defined.

It should no happen if you export the class/function

**Comment by CanerSezgin** [comment]
Hello @Betree

When I merged **master** into **branch** I could get rid of the other changes (when I run rebase, it hadn't worked as you remember) It is kinda ready now and I think you can review all. 

I left a TODO inside code, I just need you to try on slack (I couldn't send request on dev mode) and also you need to set up correct webhook for it. 

Example messages that I create is below. I have changed little bit of the flow. Such as if there is a bad word more than one property (name, description, website etc) I am returning a message it says this property includes these bad words. (I use **\n** inside string but not sure how slack is handling this)

>    Field "name" for <qketoed lsfpills |http://localhost:3000/qketoed-lsfpills> was set to a suspicious value/s: [ keto, pills ]
>     Field "website" for <qketoed lsfpills |http://localhost:3000/qketoed-lsfpills> was set to a suspicious value/s: [ keto, pills ]
>     Field "description" for <qketoed lsfpills |http://localhost:3000/qketoed-lsfpills> was set to a suspicious value/s: [ pills ]

**Comment by Betree** [comment]
Actually there are changes that you can make before I move forward, adding a review now

File: server/models/Collective.js [hunk 1]
```diff
@@ -653,8 +654,17 @@ export default function(Sequelize, DataTypes) {
             });
           }
 
+          if (instance.type === 'COLLECTIVE') {
```
Comment (Betree) [inline server/models/Collective.js]: We should actually run that on all types of collectives (events, users, orgs...etc)

File: server/models/Collective.js [hunk 2]
```diff
@@ -653,8 +654,17 @@ export default function(Sequelize, DataTypes) {
...
+            spamController.collectiveCheck(instance);
+          }
+
           return null;
         },
+        afterUpdate: async instance => {
+          if (instance.type === 'COLLECTIVE') {
```
Comment (Betree) [inline server/models/Collective.js]: Same: we should actually run that on all types of collectives (events, users, orgs...etc)

File: server/lib/spam.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      }
+    });
+    return suspiciousWords;
+  }
+
+  /**
+   *
+   * @param collective {Collective}
+   */
+  private collectiveCheck(collective: Collective): { collective: Collective; warnings: object } {
```
Comment (Betree) [inline server/lib/spam.ts]: Good idea to return a map for `warnings`, it's a good way to bundle all of them.

Why returning the collective? You pass it as a parameter so you already have it when you call this function.

File: server/lib/spam.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    // TODO: Add WEBHOOK
+    if (Object.keys(result).length) {
+        slackLib.postActivityOnPublicChannel(
+          {
+            type: activities.COLLECTIVE_BADWORD_DETECTED,
+            data: result,
+          },
+          'WEBHOOK',
+        );
+    }
```
Comment (Betree) [inline server/lib/spam.ts]: If using an activity then you should not use `slackLib` directly, you should just create the activity

File: server/lib/spam.ts (hunk 3 above)
Comment (CanerSezgin) [reply server/lib/spam.ts]: it's my bad, the structure is a bit different than this. Then, I moved the whole logic inside spam.js and call slack function inside. Now don't need to return anything. Resolved.

File: server/lib/spam.ts (hunk 4 above)
Comment (CanerSezgin) [reply server/lib/spam.ts]: Couldn't get how to do this. You can send me an example or basically take it from here, again it is slack related (you need to add webhook etc in anyway). Not sure which functions/files should I use exactly.

**Comment by CanerSezgin** [comment]
@Betree except from this one https://github.com/opencollective/opencollective-api/pull/3119#discussion_r364689056 all requests have been sorted out. Thanks for reviewing.

**Comment by Betree** [comment]
I'll take it from there, thanks @CanerSezgin :+1:

**Pull request merged**